bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/storage/ldb/PersistentEntryLogMetadataMap.java

```diff
@@ -0,0 +1,185 @@
+        @Override
+        protected DataInputStream initialValue() {
+            return new DataInputStream(bais.get());
+        }
+    };
+
+    public PersistentEntryLogMetadataMap(String path, ServerConfiguration conf) throws IOException {
+        String metadataPath = FileSystems.getDefault().getPath(path, "metadata").toFile().toString();
+        metadataMapDB = KeyValueStorageRocksDB.factory.newKeyValueStorage(metadataPath, DbConfigType.Small, conf);
+    }
+
+    @Override
+    public boolean containsKey(long entryLogId) throws EntryLogMetadataMapException {
+        LongWrapper key = LongWrapper.get(entryLogId);
+        try {
+            boolean isExist;
+            try {
+                isExist = metadataMapDB.get(key.array) != null;
+            } catch (IOException e) {
+                throw new EntryLogMetadataMapException(e);
+            }
+            return isExist;
+        } finally {
+            key.recycle();
+        }
+    }
+
+    @Override
+    public void put(long entryLogId, EntryLogMetadata entryLogMeta) throws EntryLogMetadataMapException {
+        LongWrapper key = LongWrapper.get(entryLogId);
+        try {
+            baos.get().reset();
+            try {
+                entryLogMeta.serialize(dataos.get());
+                dataos.get().flush();
+                metadataMapDB.put(key.array, baos.get().toByteArray());
+            } catch (IllegalStateException | IOException e) {
+                LOG.error("Failed to deserialize entrylog-metadata, entryLogId {}", entryLogId);
+                throw new EntryLogMetadataMapException(e);
+            }
+        } finally {
+            key.recycle();
+        }
+
+    }
+
+    @Override
+    public void forEach(BiConsumer<Long, EntryLogMetadata> action) throws EntryLogMetadataMapException {
+        CloseableIterator<Entry<byte[], byte[]>> iterator = metadataMapDB.iterator();
+        try {
+            while (iterator.hasNext()) {
+                if (isClosed.get()) {
+                    break;
+                }
+                Entry<byte[], byte[]> entry = iterator.next();
+                long entryLogId = ArrayUtil.getLong(entry.getKey(), 0);
+                ByteArrayInputStream localBais = bais.get();
+                DataInputStream localDatais = datais.get();
+                if (localBais.available() < entry.getValue().length) {
+                    localBais.close();
+                    localDatais.close();
+                    ByteArrayInputStream newBais = new ByteArrayInputStream(entry.getValue());
+                    bais.set(newBais);
+                    datais.set(new DataInputStream(newBais));
+                } else {
+                    localBais.read(entry.getValue(), 0, entry.getValue().length);
+                }
+                localBais.reset();
+                localDatais.reset();
+                action.accept(entryLogId, EntryLogMetadata.deserialize(datais.get()));
```

Comment:
this 'forEach' call is going to be quite expensive on jvm gc call. Since it will create EntryLogMetadata object for each entrylog file. You should consider optimizing it.

Reply:
yes, I had the same concern but I forgot to make this change. but, now I have added recyclable entryMetadata class rather making any change in existing `EntryLogMetadata` because I don't want to change life-cycle semantic of existing `EntryLogMetadata`